src/libraries/System.Private.CoreLib/src/System/Threading/PortableThreadPoolEventSource.cs

```diff
@@ -3,117 +3,286 @@
+        public static class Opcodes // this name and visibility is important for EventSource
         {
-            public const EventOpcode WaitOpcode = (EventOpcode)90;
-            public const EventOpcode SampleOpcode = (EventOpcode)100;
-            public const EventOpcode AdjustmentOpcode = (EventOpcode)101;
-            public const EventOpcode StatsOpcode = (EventOpcode)102;
+            public const EventOpcode IOEnqueue = (EventOpcode)13;
+            public const EventOpcode IODequeue = (EventOpcode)14;
+            public const EventOpcode Wait = (EventOpcode)90;
+            public const EventOpcode Sample = (EventOpcode)100;
+            public const EventOpcode Adjustment = (EventOpcode)101;
+            public const EventOpcode Stats = (EventOpcode)102;
         }
 
-        public static class Keywords
+        public static class Keywords // this name and visibility is important for EventSource
         {
             public const EventKeywords ThreadingKeyword = (EventKeywords)0x10000;
+            public const EventKeywords ThreadTransferKeyword = (EventKeywords)0x80000000;
         }
 
         private PortableThreadPoolEventSource()
+            : base(
+                  new Guid(0xe13c0d23, 0xccbc, 0x4e12, 0x93, 0x1b, 0xd9, 0xcc, 0x2e, 0xee, 0x27, 0xe4),
+                  "Microsoft-Windows-DotNETRuntime")
         {
         }
 
-        [Event(1, Level = EventLevel.Informational, Message = WorkerThreadMessage, Task = Tasks.WorkerThreadTask, Opcode = EventOpcode.Start, Version = 0, Keywords = Keywords.ThreadingKeyword)]
-        public void WorkerThreadStart(short numExistingThreads)
+        [NonEvent]
+        private unsafe void WriteThreadEvent(int eventId, int numExistingThreads)
         {
-            WriteEvent(1, numExistingThreads);
+            int retiredWorkerThreadCount = 0;
+            short clrInstanceId = ClrInstanceId;
+
+            EventData* data = stackalloc EventData[3];
+            data[0].DataPointer = (IntPtr)(&numExistingThreads);
+            data[0].Size = sizeof(int);
+            data[0].Reserved = 0;
+            data[1].DataPointer = (IntPtr)(&retiredWorkerThreadCount);
+            data[1].Size = sizeof(int);
+            data[1].Reserved = 0;
+            data[2].DataPointer = (IntPtr)(&clrInstanceId);
+            data[2].Size = sizeof(short);
+            data[2].Reserved = 0;
+            WriteEventCore(eventId, 3, data);
         }
 
-        [Event(2, Level = EventLevel.Informational, Message = WorkerThreadMessage, Task = Tasks.WorkerThreadTask, Opcode = EventOpcode.Stop, Version = 0, Keywords = Keywords.ThreadingKeyword)]
-        public void WorkerThreadStop(short numExistingThreads)
+        [Event(50, Level = EventLevel.Informational, Message = Messages.WorkerThread, Task = Tasks.ThreadPoolWorkerThread, Opcode = EventOpcode.Start, Version = 0, Keywords = Keywords.ThreadingKeyword)]
+        public unsafe void ThreadPoolWorkerThreadStart(int numExistingThreads)
```

Comment:
Most of these events, as implemented, aren't going to work.  They will be emitted properly, but won't really be consumable in a supported way.  The reason is that the method signature needs to match the event arguments.  When the EventSource is instantiated, it creates a manifest by walking the set of event methods, and building up their metadata through reflection over the method arguments.  Then, the manifest is emitted into the event stream so that tools can consume it and have knowledge of how to decode the events.  As an example, in the case of ThreadPoolWorkerThreadStart, the EventSource generated manifest will state that the only payload argument is an int called numExistingThreads.  In reality, you're actually writing out 3 payload arguments.

I would recommend simplifying this code a bit, and using the standard EventSource pattern of:

```
[Event(id, ...)]
public void EventName(payloadArgs...)
{
    WriteEvent(id, payloadArgs...);
}
```

This should be all that you need.  You also shouldn't need to call IsEnabled inside of the event method - this will happen inside of WriteEvent, and at this point, you've already computed the arguments.  If you do need to compute expensive arguments, you can use a [NonEvent] method as an outer wrapper around an event method to make sure that you only compute the arguments if the event is enabled.

Totally fine to keep all of the existing Keyword, OpCode, Task and Version stuff, though Version=0 is the default.

Reply:
Thanks, fixed. Due to the ClrInstanceId being ushort, I couldn't use any existing overload of WriteEvent().